File: SLAM_zero_to_hero/1_16/ros1/src/slam_tutorial/scripts/subscriber.py
#!/usr/bin/env python3
"""
ROS1 Subscriber Example

Subscribes to robot pose messages.
Demonstrates basic ROS1 Python subscriber pattern.
"""

import rospy
from geometry_msgs.msg import PoseStamped
import math


class PoseSubscriber:
    def __init__(self):
        # Initialize node
        rospy.init_node('pose_subscriber', anonymous=True)

        # Create subscriber
        self.sub = rospy.Subscriber('/robot/pose', PoseStamped, self.callback)

        # Store last position for distance calculation
        self.last_x = None
        self.last_y = None
        self.total_distance = 0.0

        rospy.loginfo("Pose subscriber started")

    def callback(self, msg):
        """Callback function for pose messages."""
        x = msg.pose.position.x
        y = msg.pose.position.y

        # Calculate distance traveled
        if self.last_x is not None:
            dx = x - self.last_x
            dy = y - self.last_y
            distance = math.sqrt(dx * dx + dy * dy)
            self.total_distance += distance

        self.last_x = x
        self.last_y = y

        # Log periodically
        if int(rospy.Time.now().to_sec() * 10) % 50 == 0:
            rospy.loginfo(
                f"Received pose: ({x:.2f}, {y:.2f}), "
                f"total distance: {self.total_distance:.2f}m"
            )

    def run(self):
        """Keep node running."""
        rospy.spin()


def main():
    subscriber = PoseSubscriber()
    subscriber.run()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: SLAM_zero_to_hero/ch02_09/ros1/src/slam_tutorial/src/subscriber.cpp
/**
 * ROS1 C++ Subscriber Example
 *
 * Subscribes to robot pose messages and tracks distance traveled.
 */

#include <ros/ros.h>
#include <geometry_msgs/PoseStamped.h>
#include <cmath>

class PoseSubscriber {
public:
    PoseSubscriber() : last_x_(0.0), last_y_(0.0), total_distance_(0.0), first_msg_(true) {
        // Subscribe to pose topic
        sub_ = nh_.subscribe("/robot/pose", 10, &PoseSubscriber::callback, this);
        ROS_INFO("C++ Pose subscriber started");
    }

    void callback(const geometry_msgs::PoseStamped::ConstPtr& msg) {
        double x = msg->pose.position.x;
        double y = msg->pose.position.y;

        // Calculate distance traveled
        if (!first_msg_) {
            double dx = x - last_x_;
            double dy = y - last_y_;
            double distance = std::sqrt(dx * dx + dy * dy);
            total_distance_ += distance;
        }
        first_msg_ = false;

        last_x_ = x;
        last_y_ = y;

        // Log periodically
        static int count = 0;
        if (++count % 50 == 0) {
            ROS_INFO("Received pose: (%.2f, %.2f), total distance: %.2fm",
                     x, y, total_distance_);
        }
    }

private:
    ros::NodeHandle nh_;
    ros::Subscriber sub_;
    double last_x_, last_y_;
    double total_distance_;
    bool first_msg_;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "pose_subscriber_cpp");

    PoseSubscriber subscriber;
    ros::spin();

    return 0;
}
